lib/patch-by-id.js

```diff
@@ -0,0 +1,187 @@
+  }
+  assert(arguments.length >= 2, 'At least two arguments are required');
+
+  var Model = this;
+  var where = {};
+  var strict = this.settings.strict;
+  // TODO: the methods which we have imported from dao (idName(),
+  // stillConnecting(), isWhereByGivenId(),...), may need to be
+  // moved to a more appropriate file (utils.js?)
+  var idPropertyName = dao.idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  // Make sure id cannot be changed
+  checkId(data[idPropertyName], id, idPropertyName, cb);
+
+  // Question[1]:
+  // Please see QUESTION-test[1] in`test\dao.suit\patch-by-id.suits.js` for failing test
+  // data = data.toObject();
+  // data = removeUndefined(data);
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    where: where,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+    var isOriginalQuery = dao.isWhereByGivenId(Model, ctx.where, id);
+
+    if (!isOriginalQuery) {
+      var err = new Error('ctx.where:' + JSON.stringify(ctx.where) + ' is not valid; the `' +
+        idPropertyName + '` original value is ' + id);
+      err.statusCode = 403;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+
+    if (strict) {
+      var instInfo = { '__unknownProperties': [] };
+      dao.applyStrictCheck(Model, strict, data,
+        instInfo, validateAndCallConnector);
+    } else {
+      validateAndCallConnector();
+    }
+
+    function validateAndCallConnector(err) {
+      if (err) return cb(err);
+      if (options.validate === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // only when options.validate is not set, take model-setting into consideration
+      if (options.validate === undefined && Model.settings.automaticValidation === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // Make sure id is not changed in the context of `before save`
+      checkId(ctx.data[idPropertyName], id, idPropertyName, cb);
+
+      dataObj = new Model(ctx.data);
+      dataObj.unsetAttribute(idPropertyName);
+      // validation required
+      dataObj.isValid(function(valid) {
```

Comment:
Yea I noticed that restriction too and unfortunately I don't think if there is any efficient alternative either... The only alternative I can think of is calling `findById`, which definitely does not make sense to call another DB query and it makes `patchById` more computational expensive... On the other hand skipping the validation makes `data` inconsistent and as you mentioned I don't think if that's a rational decision either.

Initially the decision for `loopback 3.0` was to map `PUT` to static `replaceById` and `PATCH` to static `patchById` instead of `prototype.updateAttributes`; however now that we cannot have the right validation for static `patchById` as a result of not having access to `instance` I agree with you that we should map `PUT` to static `replaceById` and `PATCH` to `prototype.updateAttributes` and we can say implementation of static `patchById` is out of scope of this milestone.